src/org/labkey/test/util/TestDataGenerator.java

```diff
@@ -257,12 +257,17 @@ private Map<String, Object> generateRow()
         final Map<String, Object> newRow = new CaseInsensitiveHashMap<>();
         for (String columnName : _columns.keySet())
         {
-            if (_autoGeneratedFields.contains(columnName) || _dataSuppliers.containsKey(columnName))
+            if (!_dataSuppliers.containsKey(columnName))
+            {
+                _dataSuppliers.put(columnName, getDefaultDataSupplier(_columns.get(columnName).getRangeURI()));
+            }
+
+            if (_autoGeneratedFields.contains(columnName))
             {
                 continue;
             }
-            // Generate default values
-            Object columnValue = getDefaultDataSupplier(_columns.get(columnName).getRangeURI()).apply(_rows);
+            // Generate values
+            Object columnValue = _dataSuppliers.get(columnName).apply(_rows);
```

Comment:
Consider making `_dataSuppliers` actually be `Suppliers`. We don't actually use the row collection in any existing data suppliers and I don't know that it adds much utility to provide it.
```
Map<String, Supplier<Object>> _dataSuppliers = new CaseInsensitiveHashMap<>()
```
Then this simplifies to:
```suggestion
            // Generate values
            Object columnValue = _dataSuppliers.get(columnName).get();
```
And we could remove the unused method `addDataSupplier(String columnName, Function<List<Map<String, Object>>, Object> supplier)`

Reply:
Thanks, I like this suggestion a lot.  (applying it to _rows without really understanding why has bothered me, this gets me out of that)
Also, this makes everything else that interacts with `_dataSuppliers `a lot more readable